params/superchain.go

```diff
@@ -210,6 +210,24 @@ func (p ProtocolVersion) Compare(other ProtocolVersion) (cmp ProtocolVersionComp
 		}
 		return outdated
 	}
+	if aPreRelease != 0 { // if A is a pre-release, then decrement the version before comparison
+		if aPatch != 0 {
+			aPatch -= 1
+		} else if aMinor != 0 {
+			aMinor -= 1
+		} else if aMajor != 0 {
+			aMajor -= 1
+		}
```

Comment:
Doesn't this mean that `1.0.0-1` winds up being `0.0.0` which is incorrectly before `0.9.0`? or `0.99999.0` etc.

I'm not entirely sure I have the logic right but I'm thinking of a test case like:

```
		{
			// Release versions in order:
			// 4.5.0 -- A is here
			// 4.9.0
			// 5.0.0-1 -- B is here
			// 5.0.0
			A:   HumanProtocolVersion{0, 4, 5, 0, 0, [8]byte{}},
			B:   HumanProtocolVersion{0, 5, 0, 0, 1, [8]byte{}},
			Cmp: OutdatedMajor,
		},
```

My thinking is that if `B` is the required version of `5.0.0-1` then we should meet that requirement (be ahead) with `5.0.0`, `5.0.0-1`, `5.0.0-2`, `5.0.1` and onwards, but we should be behind it if we have `4.` anything.

Reply:
good catch, after decrementing the lowest non-zero component we need to set the component lower than that to the max value. E.g. `1.0.0-1` becomes `0.MAX.0`, which is then higher than `0.9.0` by minor-level.